joblib/func_inspect.py

```diff
@@ -127,28 +127,10 @@ def get_func_name(func, resolv_alias=True, win_characters=True):
         if filename is not None:
             # mangling of full path to filename
             parts = filename.split(os.sep)
-            if parts[-1].startswith('<ipython-input'):
-                # We're in a IPython (or notebook) session. parts[-1] comes
-                # from func.__code__.co_filename and is of the form
-                # <ipython-input-N-XYZ>, where:
-                # - N is the cell number where the function was defined
-                # - XYZ is a hash representing the function's code (and name).
-                #   It will be consistent across sessions and kernel restarts,
-                #   and will change if the function's code/name changes
-                # We remove N so that cache is properly hit if the cell where
-                # the func is defined is re-exectuted.
-                # The XYZ hash should avoid collisions between functions with
-                # the same name, both within the same notebook but also across
-                # notebooks
-                splitted = parts[-1].split('-')
-                parts[-1] = '-'.join(splitted[:2] + splitted[3:])
-            elif len(parts) > 2 and parts[-2].startswith('ipykernel_'):
-                # In a notebook session (ipykernel). Filename seems to be 'xyz'
-                # of above. parts[-2] has the structure ipykernel_XXXXXX where
-                # XXXXXX is a six-digit number identifying the current run (?).
-                # If we split it off, the function again has the same
-                # identifier across runs.
-                parts[-2] = 'ipykernel'
+            is_jupyter = (parts[-1].startswith('<ipython-input') or
+                          (len(parts) > 2 and parts[-2].startswith('ipykernel_')))
+            if is_jupyter:
+                parts = os.getcwd().split(os.sep) + ["<ipython-input>"]
```

Comment:
I think I prefer that the first part would stay `__main__` as previously 
to mark the fact that it's interactively defined code.

Furthermore we should consolidate what we have learned about ipykernel scoping.

```suggestion
                # Source filenames generated by ipykernel sessions from jupyter
                # do not give enough information to reconstruct the name
                # of the original ".ipynb" source file so we have to scope such
                # functions in a jupyter specific manner.
                #
                # Generate a synthetic function scope that reflects that:
                # - it's an interactively defined function (__main__),
                # - it comes from an IPython / Jupyter session
                # - this session is scoped by an identifier reflecting the current
                #   working directory of the ipykernel process, which is typically
                #   the folder of the jupyter notebook (but not 100% guaranteed.
                #
                # This way the scope is stable across jupyter restarts or edits
                # made to the cell content that holds the function and does not
                # collide with the scope of function defined in regular Python
                # source files.
                #
                # There is the possibility that this scope will collide with an
                # identically named function defined in a notebook executing code
                # in the working directory but this is a fundamental limitation
                # of the lack of information provided by the ipykernel context.
                #
                # In practice this means that we can trigger spurious cache
                # invalidations from concurrently running notebook in the same
                # folder with colliding interactively defined function names.
                parts = (
                    ["__main__"] + os.getcwd().split(os.sep) + ["<ipython-input>"]
                )
```

Reply:
also I am not sure if you my code suggest is 79 chars max.